cpp/src/parquet/encryption/key_toolkit.cc

```diff
@@ -35,6 +46,89 @@ std::shared_ptr<KmsClient> KeyToolkit::GetKmsClient(
       });
 }
 
+// Filter out files that are not suitable for key rotation.
+inline bool filter_out_file(std::string child) {
+  // Filters out hidden files. A file is considered to
+  // be hidden, and should not be considered for processing, when the file name
+  // starts with a period ('.') or an underscore ('_').
+  if (child.at(0) == kPeriod || child.at(0) == kUnderscore) return true;
+  return false;
+}
+
+void KeyToolkit::RotateMasterKeys(const KmsConnectionConfig& kms_connection_config,
+                                  const std::shared_ptr<FilePath>& folder_path,
+                                  bool double_wrapping, double cache_lifetime_seconds) {
+  // If process wrote files with double-wrapped keys, clean KEK cache (since master keys
+  // are changing). Only once for each key rotation cycle; not for every folder
+  const auto now = internal::CurrentTimePoint();
+  auto lock = last_cache_clean_for_key_rotation_time_mutex_.Lock();
+  if (now > last_cache_clean_for_key_rotation_time_ +
+                std::chrono::duration<double>(kCacheCleanPeriodForKeyRotation)) {
+    kek_write_cache_per_token().Clear();
+    last_cache_clean_for_key_rotation_time_ = now;
+  }
+  lock.Unlock();
+  std::vector<::arrow::fs::FileInfo> parquet_files_in_folder;
+  ::arrow::fs::FileSelector s;
+  s.base_dir = folder_path->dir_name();
+  parquet_files_in_folder = folder_path->filesystem()->GetFileInfo(s).ValueOrDie();
+
+  if (parquet_files_in_folder.size() == 0) {
+    throw ParquetException("Couldn't rotate keys - no parquet files in folder " +
+                           folder_path->dir_name());
+  }
+
+  for (auto const& parquet_file : parquet_files_in_folder) {
+    if (parquet_file.type() != ::arrow::fs::FileType::File)
+      throw ParquetException("Expecting file type in " + folder_path->dir_name());
+    std::string child = parquet_file.base_name();
+    if (filter_out_file(child)) continue;
```

Comment:
Would it make more sense to only look for files that begin with the `_KEY_MATERIAL_FOR_` prefix, rather than assuming that any file that doesn't begin with "." or "_" is a Parquet file with an associated external key material file? Otherwise this will fail for any directories that contain any other type of file.